File: app/src/main/java/com/example/messenger/UserViewModel.java
package com.example.messenger;


import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public class UserViewModel extends ViewModel {

    private final FirebaseAuth auth;
    private final FirebaseDatabase database;
    private final DatabaseReference userReference;

    private final MutableLiveData<FirebaseUser> user = new MutableLiveData<>();
    private final MutableLiveData<ArrayList<User>> users = new MutableLiveData<>();

    public UserViewModel() {
        database = FirebaseDatabase.getInstance();
        userReference = database.getReference("Users");
        auth = FirebaseAuth.getInstance();
        auth.addAuthStateListener(firebaseAuth -> user.setValue(firebaseAuth.getCurrentUser()));
        userReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                ArrayList<User> userList = new ArrayList<>();
                FirebaseUser currentUser = auth.getCurrentUser();
                if(currentUser == null) {
                    return;
                }

                for (DataSnapshot dataSnapshot:
                     snapshot.getChildren()) {
                    User user = dataSnapshot.getValue(User.class);
                    if(user == null) {
                        return;
                    }
//                    if(!user.getId().equals(currentUser.getUid())) {
                        userList.add(user);
//                    }
                }
                users.setValue(userList);
            }
            @Override
            public void onCancelled(@NonNull DatabaseError error) {}
        });
    }

    public LiveData<FirebaseUser> getUser() {
        return user;
    }

    public LiveData<ArrayList<User>> getUsers() {
        return users;
    }

    public void logout() {
        auth.signOut();
        onlineUserStatus(false);
    }

    public void onlineUserStatus(boolean isOnline) {
        FirebaseUser firebaseUser = auth.getCurrentUser();
        if(firebaseUser == null) {
            return;
        }
        userReference.child(firebaseUser.getUid()).child("online").setValue(isOnline);
    }
}

File: app/src/main/java/com/example/messenger/UserViewModel.kt
package com.example.messenger

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.messenger.pojo.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener


class UserViewModel : ViewModel() {
    private val user = MutableLiveData<FirebaseUser>()
    private val users = MutableLiveData<List<User>>()
    private val auth: FirebaseAuth = FirebaseAuth.getInstance()
    private val database: FirebaseDatabase = FirebaseDatabase.getInstance()
    private val userReference: DatabaseReference = database.getReference("Users")

    init {
        auth.addAuthStateListener { user.value = it.currentUser }
        userReference.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val currentUser = auth.currentUser ?: return
                val userList = mutableListOf<User>()
                for (item in snapshot.children) {
                    val user = item.getValue(User::class.java) ?: return
                    if(currentUser.uid != user.id) userList.add(user)
                }
                users.value = userList
            }
            override fun onCancelled(error: DatabaseError) {}
        })
    }

    val getUser: LiveData<FirebaseUser> get() = user
    val getUsers: LiveData<List<User>> get() = users

    fun logout() {
        onlineUserStatus(false)
        auth.signOut()
    }

    fun onlineUserStatus(isOnline: Boolean) {
        userReference
            .child(auth.currentUser?.uid ?: return)
            .child("online")
            .setValue(isOnline)
    }
}